File: src/main/java/gregtechlite/gtlitecore/core/network/PacketHandler.java
package gregtechlite.gtlitecore.core.network;

import gregtechlite.gtlitecore.api.network.NetworkPacket;
import net.minecraft.util.IntIdentityHashBiMap;
import org.jetbrains.annotations.ApiStatus.Internal;

@Internal
public class PacketHandler
{

    private static final PacketHandler INSTANCE = new PacketHandler(10);

    private final IntIdentityHashBiMap<Class<? extends NetworkPacket>> packets;

    private int id = 1;

    private PacketHandler(int initialCapacity)
    {
        this.packets = new IntIdentityHashBiMap<>(initialCapacity);
    }

    /* package */ static PacketHandler getInstance()
    {
        return INSTANCE;
    }

    public void registerPacket(Class<? extends NetworkPacket> packetClass)
    {
        this.packets.put(packetClass, id++);
    }

    public int getPacketId(Class<? extends NetworkPacket> packetClass)
    {
        return this.packets.getId(packetClass);
    }

    public Class<? extends NetworkPacket> getPacketClass(int packetId)
    {
        return this.packets.get(packetId);
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/core/network/PacketHandler.kt
package gregtechlite.gtlitecore.core.network

import gregtechlite.gtlitecore.api.collection.intIdHashBiMapOf
import gregtechlite.gtlitecore.api.network.NetworkPacket

class PacketHandler(initialCapacity: Int = 10)
{
    private val packets = intIdHashBiMapOf<Class<out NetworkPacket>>(initialCapacity)
    private var id = 1

    companion object
    {
        private val instance = PacketHandler()

        fun getInstance(): PacketHandler = instance
    }

    fun registerPacket(packetClass: Class<out NetworkPacket>)
    {
        packets.put(packetClass, id++)
    }

    fun getPacketId(packetClass: Class<out NetworkPacket>) = packets.getId(packetClass)

    fun getPacketClass(packetId: Int): Class<out NetworkPacket> = packets[packetId]!!
}